File: spiralNumbers/spiralNumbers.js
/*

Construct a square matrix with a size N × N containing integers from 1 to N * N in a spiral order, starting from top-left and in clockwise direction.

Example

For n = 3, the output should be

solution(n) = [[1, 2, 3],
               [8, 9, 4],
               [7, 6, 5]]
Input/Output

[execution time limit] 4 seconds (js)

[memory limit] 1 GB

[input] integer n

Matrix size, a positive integer.

Guaranteed constraints:
3 ≤ n ≤ 100.

[output] array.array.integer

*/

function printMatrix(matrix)
{
    let tmp = "";

    for (let i = 0; i < matrix.length; i++)
    {
        tmp = "";

        for (let j = 0; j < matrix[0].length; j++)
        {
            tmp = tmp + matrix[i][j] + "\t";
        }

        console.log(tmp);

    }

}

function createNxNArray(n)
{
    let m = new Array();
    
    for (let i = 0; i < n; i++)
    {
        m [i] = new Array();
        
        for (let j = 0; j < n; j++)
        {
            m [i] [j] = "x";
        }

    }

    // printMatrix(m);

    return m;
}

function spiralNumbers(n)
{
    let m = createNxNArray(n); 
    let cnt = 1;
    let rightBorderX = 0;
    let rightBorderY = m.length;
    let downBorderX = m.length - 1;
    let downBorderY = 0;
    let leftBorderX = m.length - 1;
    let leftBorderY = 0;
    let upBorderX = m.length - 1;
    let upBorderY = 0;

    // for (let i = 0; i < Math.pow(m.length, 2); i++)
    while (cnt <= Math.pow(m.length, 2) && cnt > 0)
    {
        // console.log(cnt);       
        
        cnt = moveRight(m, cnt, rightBorderX, rightBorderY);
        rightBorderX++;
        rightBorderY--;
        
        if (cnt <= Math.pow(m.length, 2) && downBorderX < m.length && downBorderY >= 0)
        {
            // console.log(i);
            
            cnt = moveDown(m, cnt - 1, downBorderX, downBorderY);
            downBorderX++;
            downBorderY--;
        }

        else
        {
            
        }

        if (cnt <= Math.pow(m.length, 2)) // && leftBorderX >= 0 && leftBorderX < m.length)
        {
            cnt = moveLeft(m, cnt - 1, leftBorderX, leftBorderY);
            leftBorderX--;
            leftBorderY++;
        }

        else
        {

        }

        if (cnt <= Math.pow(m.length, 2)) // && upBorderX >= 0 && upBorderY < m.length)
        {
            cnt = moveUp(m, cnt - 1, upBorderX, upBorderY);
            upBorderX--;
            upBorderY++;
        }

        else
        {

        }

    }
    
    printMatrix(m);
    // return m;

}

function moveRight(a, cnt, rightBorderX, rightBorderY)
{

    for (let i = rightBorderX; i < rightBorderY; i++)
    {
        a [rightBorderX] [i] = cnt++;
    }

    return cnt;
}

function moveDown(a, cnt, downBorderX, downBorderY)
{

    for (let i = downBorderY; i < downBorderX; i++)
    {
        // console.log("i:", i, "cnt:", cnt, "downBorderX:", downBorderX, "downBorderY:", downBorderY);
        a [i] [downBorderX] = cnt;
        cnt++;
    }

    return cnt;
}

function moveLeft(a, cnt, leftBorderX, leftBorderY)
{

    for (let i = leftBorderX; i >= leftBorderY; i--)
    {
        a [leftBorderX] [i] = cnt;
        cnt++;
    }

    return cnt;
}

function moveUp(a, cnt, upBorderX, upBorderY)
{

    for (let i = upBorderX; i > upBorderY; i--)
    {
        a [i] [upBorderY] = cnt;
        cnt++;
    }

    return cnt;
}

function solution(n) 
{
    return spiralNumbers(n);
}

let n = 3;  // the output should be solution(n) =  [[1, 2, 3],
//                                                  [8, 9, 4],
//                                                  [7, 6, 5]]


spiralNumbers(3);

/*



********
BONEYARD
********

var spiral =    [moveRight(m, 1, rightBorderX, rightBorderY), 
        moveDown(m, cnt - 1, downBorderX, downBorderY), 
        moveLeft(m, cnt - 1, leftBorderX, leftBorderY),
        moveUp(m, cnt - 1, upBorderX, upBorderY)];
    spiral[0](m, 1, rightBorderX, rightBorderY);

function moveRight(a, cnt)
{

    for (let i = 0; i < a.length; i++)
    {
        a [0] [i] = cnt++;
    }

    return cnt;
}

function moveDown(a, cnt)
{

    for (let i = 0; i < a.length; i++)
    {
        a [i] [a.length - 1] = cnt;
        cnt++;
    }

    return cnt;
}

function moveLeft(a, cnt)
{

    for (let i = a.length - 1; i >= 0; i--)
    {
        console.log(i, cnt);
        a [a.length - 1] [i] = cnt;
        cnt++;
    }

    return cnt;
}

function moveUp(a, cnt)
{

    for (let i = a.length - 1; i > 0; i--)
    {
        a [i] [0] = cnt;
        cnt++;
    }

    return cnt;
}

console.log(i, cnt);
        
// createNxNArray(n);
// console.log(i, cnt);
// console.log("$");
        // console.log("i:", i, "r:", r, "c:", c);
    // console.log("$");
        // console.log("i:", i, "r:", r, "c:", c);
// console.log("!", m);
    // console.log("@", m);
    // m = moveDown(m, 0, 3, 4);
    // console.log("#");
        // console.log("i:", i, "r:", r, "c:", c);
    // console.log(a);
    // console.log("#");

*/

File: spiralNumbers/spiral_numbers.py
"""

Construct a square matrix with a size N × N containing integers from 1 to N * N in a spiral order, starting from top-left and in clockwise direction.

Example

For n = 3, the output should be

solution(n) = [[1, 2, 3],
               [8, 9, 4],
               [7, 6, 5]]
Input/Output

[execution time limit] 4 seconds (js)

[memory limit] 1 GB

[input] integer n

Matrix size, a positive integer.

Guaranteed constraints:
3 ≤ n ≤ 100.

[output] array.array.integer

"""

import math

def print_matrix(matrix) :
# {
    tmp = ""

    for i in len(matrix[0]) :
    # {
        tmp = tmp + matrix[i][j] + "\t"
    # }

    print(tmp)

# }

def create_NxN_array(n) :
# {
    m = [None] * n

    for i in range(n) :
    # {
        # print(i)
        m[i] = [None] * n

        for j in range(n) :
        # { 
            m[i][j] = 'x'
        # }

    # }

    return m

# }

def spiral_numbers(n) :
# {
    m = create_NxN_array(n)
    cnt = 1
    right_borderX = 0
    right_borderY = len(m)
    down_borderX = 0
    down_borderY = len(m)
    left_borderX = len(m) - 1
    left_borderY = 0
    up_borderX = len(m) - 1
    up_borderY = 0

    while(cnt <= math.pow(len(m), 2) and cnt > 0) :
    # {        
        print("!", cnt, m)

        cnt = move_right(m, cnt, right_borderX, right_borderY)
        right_borderX = right_borderX + 1
        right_borderY = right_borderY - 1

        if (cnt <= math.pow(len(m), 2)) :
        # {
            cnt = move_down(m, cnt - 1, down_borderX, down_borderY)
            down_borderX = down_borderX + 1
            down_borderY = down_borderY - 1
            print("@", cnt, m)
        # }

        else :
        # {
            None
        # }

        if (cnt <= math.pow(len(m), 2)) :
        # {
            
            cnt = move_left(m, cnt - 1, left_borderX, left_borderY)
            left_borderX = left_borderX - 1
            left_borderY = left_borderY + 1
            print("#", cnt, m)

        # }

        else :
        # {
            None
        # }

        if (cnt <= math.pow(len(m), 2)) :
        # {
            
            cnt = move_up(m, cnt - 1, up_borderX, up_borderY)
            up_borderX = up_borderX - 1
            up_borderY = up_borderY + 1
            print("$", cnt, m)

        # }

        else :
        # {
            None
        # }

    # }

    return m

# }

def move_right(a, cnt, right_borderX, right_borderY) :
# {

    # for (let i = rightBorderX; i < rightBorderY; i++)
    for i in range(right_borderX, right_borderY) :
    # {
        # console.log("! i:", i, "cnt:", cnt, "rightBorderX:", rightBorderX, "rightBorderY:", rightBorderY);
        a[right_borderX][i] = cnt
        cnt = cnt + 1
    # }

    return cnt
# }

def move_down(a, cnt, downBorderX, downBorderY) :
# {

    # for (let i = downBorderX; i < downBorderY; i++)
    for i in range(downBorderX, downBorderY) :
    # {
        # console.log("@ i:", i, "cnt:", cnt, "downBorderX:", downBorderX, "downBorderY:", downBorderY);
        a[i][downBorderY - 1] = cnt
        cnt = cnt + 1
    # }

    return cnt
# }

def move_left(a, cnt, left_borderX, left_borderY) :
# {
    # for (let i = leftBorderX; i >= leftBorderY; i--) :
    # for i in range(left_borderX, left_borderY) :
    for i in range(left_borderY, left_borderX, -1) :
    # {
        # print(a, " # i:", i, "cnt:", cnt, "leftBorderX:", left_borderX, "leftBorderY:", left_borderY)
        a[left_borderX][i] = cnt
        cnt = cnt + 1
    # }

    return cnt
# }

def move_up(a, cnt, up_borderX, up_borderY) :
# {
    # for (let i = upBorderX; i > upBorderY; i--) 
    # for i in range(up_borderX, up_borderY) :
    for i in range(up_borderY, up_borderX, -1) :
    # {
        # print("$ i:", i, "cnt:", cnt, "up_borderX:", up_borderX, "up_borderY:", up_borderY)
        a[i][up_borderY] = cnt
        cnt = cnt + 1
    # }

    return cnt
# }

def solution(n) :
# {
    return spiral_numbers(n)
# }    

n = 3;   #the output should be solution(n) =  [[1, 2, 3],
#                                              [8, 9, 4],
#                                              [7, 6, 5]]

# spiral_numbers(5)
print(solution(n))

"""

********
BONEYARD
********

if (up_borderX <= 0) :
            # {
                None
            # }

            else :
            # {
                None
            # }
            
            if (up_borderY >= len(m)) :
            # {
                None
            # }    
                
            else :
            # {
                None # check if reaches len(m)
            # }

# let downBorderX = 0;
        # let downBorderY = m.length;
        

            if (left_borderX <= 0) :
            # {
                None 
            # }

            else :
            # {
                None
            # }
            
            if (left_borderY >= len(m)) :
            # {
                None
            # }

            else :
            # {
                None # check if reaches len(m)
            # }

            # print("!", "cnt:", cnt, "leftBorderX:", left_borderX, "leftBorderY:", left_borderY)
            # print("@", "cnt:", cnt, "leftBorderX:", left_borderX, "leftBorderY:", left_borderY)
            # print(cnt, right_borderX, right_borderY, down_borderX, down_borderY, left_borderX, left_borderY, up_borderX, up_borderY)

Creating an empty list:

>>> xs = [None] * 10
>>> xs
[None, None, None, None, None, None, None, None, None, None]
Assigning a value to an existing element of the above list:

>>> xs[1] = 5
>>> xs
[None, 5, None, None, None, None, None, None, None, None]

"""